mqtt/mqtt-broker/src/configuration.rs

```diff
@@ -0,0 +1,304 @@
+use regex::Regex;
+use serde::{Deserialize, Deserializer};
+use std::path::Path;
+use std::time::Duration;
+
+pub const DEFAULTS: &str = include_str!("../config/default.json");
+
+#[derive(Debug, Deserialize)]
+pub enum QueueFullAction {
+    #[serde(rename(deserialize = "drop_new"))]
+    DropNew,
+
+    #[serde(rename(deserialize = "drop_old"))]
+    DropOld,
+
+    #[serde(rename(deserialize = "disconnect"))]
+    Disconnect,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct InflightMessages {
+    max_count: u32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct RetainedMessages {
+    max_count: u32,
+
+    #[serde(with = "humantime_serde")]
+    expiration: Duration,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionMessages {
+    #[serde(deserialize_with = "humansize")]
+    max_message_size: u64,
+
+    max_count: u32,
+
+    #[serde(deserialize_with = "humansize")]
+    max_total_space: u64,
+
+    when_full: QueueFullAction,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct SessionPersistence {
+    file_path: String,
+
+    #[serde(with = "humantime_serde")]
+    time_interval: Duration,
+    unsaved_message_count: u32,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Session {
+    #[serde(with = "humantime_serde")]
+    expiration: Duration,
+    messages: SessionMessages,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BrokerConfig {
+    inflight_messages: InflightMessages,
+    retained_messages: RetainedMessages,
+    session: Session,
+    persistence: Option<SessionPersistence>,
+}
+
+pub fn humansize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where
+    T: std::str::FromStr + std::ops::Mul<Output = T> + std::convert::From<u32>,
+    D: Deserializer<'de>,
+{
+    lazy_static! {
+        static ref SIZE_PATTERN: Regex = Regex::new(r"\s*(\d+)\s*([a-zA-Z]+)\s*")
+            .expect("failed to create new Regex from pattern");
+    }
+
+    let s = String::deserialize(deserializer)?;
+
+    let captures = SIZE_PATTERN
+        .captures(&s.as_str())
+        .ok_or_else(|| error::<D>(&s, &"256kb"))?;
+    let base = captures[1]
+        .parse::<T>()
+        .or_else(|_| Err(error::<D>(&captures[1], &"256")))?;
+
+    let multiplier = captures[2].to_lowercase();
+    let multiplier = get_multiplier::<T, D>(multiplier.as_ref())?;
+
+    Ok(base * multiplier)
+}
+
+fn get_multiplier<'de, T, D>(str: &str) -> Result<T, D::Error>
+where
+    T: std::convert::From<u32>,
+    D: Deserializer<'de>,
+{
+    let result = match str {
+        "b" => 1,
+        "kb" => 1024,
+        "mb" => 1024 * 1024,
+        "gb" => 1024 * 1024 * 1024,
+        _ => return Err(error::<D>(str, &"'b', 'kb', 'mb' or 'gb'")),
+    };
+
+    Ok(result.into())
+}
+
+fn error<'de, D>(unexpected: &str, expected: &str) -> D::Error
+where
+    D: Deserializer<'de>,
+{
+    serde::de::Error::invalid_value(serde::de::Unexpected::Str(&unexpected), &expected)
+}
+
+impl BrokerConfig {
+    pub fn new() -> Result<Self, ConfigError> {
```

Comment:
In what case we use this method and not from_file?

Reply:
it was @myagley suggestion, see a bit above, also how to use it when no config file is defined to override defaults.

Reply:
Ah, just found it. Still curious why it is better :)